dataset/src/main/scala/frameless/TypedDataset.scala

```diff
@@ -151,6 +151,48 @@ class TypedDataset[T] protected[frameless](val dataset: Dataset[T])(implicit val
     */
   def project[U](implicit projector: SmartProject[T,U]): TypedDataset[U] = projector.apply(this)
 
+  /** Returns a new [[TypedDataset]] that contains the elements of both this and the `other` [[TypedDataset]]
+    * combined.
+    *
+    * Note that, this function is not a typical set union operation, in that it does not eliminate
+    * duplicate items. As such, it is analogous to `UNION ALL` in SQL.
+    *
+    * Differs from `Dataset#union` by aligning fields if possible.
+    * It will not compile if `Datasets` have not compatible schema.
+    *
+    * Example:
+    * {{{
+    *   case class Foo(x: Int, y: Long)
+    *   case class Bar(y: Long, x: Int)
+    *   case class Faz(x: Int, y: Int, z: Int)
+    *
+    *   df1: TypedDataset[Foo] = ...
+    *   df2: TypedDataset[Bar] = ...
+    *   df3: TypedDataset[Faz] = ...
+    *
+    *   df1.union(df2): TypedDataset[Foo]
+    *
+    *   // fails to compile
+    *   df1.union(df3)
```

Comment:
I am not really sure why Foo and Bar can be unioned but Foo and Faz should not. Do you have a use-case where one wouldn't want this to compile?

Reply:
I don't have strong opinion on this. I just went with more "rigorous" behavior but when I think of it I actually agree - when doing union I usually have the first Dataset in mind. Should it use schema from first Dataset and discard extra fields from the second? 

That's what I have in mind:
```scala
case class Foo[A, B](a: A, b: B)
case class Bar[A, B](b: B, a: A)
case class Baz[A, B, C](a: A, b: B, c: C)

foo union bar // compiles
bar union foo // compiles
foo union baz // compiles, "discards" C
baz union foo // it fails with SmartProject but I guess it should compile too?
```

Reply:
Yes, smart project will only work going from more fields down to less. It will not compile if you try to go from less to more.

Reply:
In practise, I guess you can always change the order of the fields to make it work.